src/metabase/lib/metric/basics.cljc

```diff
@@ -0,0 +1,27 @@
+(ns metabase.lib.metric.basics
+  (:require
+   [metabase.lib.equality :as lib.equality]
+   [metabase.lib.join :as-alias lib.join]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.lib.metadata.calculation :as lib.metadata.calculation]
+   [metabase.lib.query :as lib.query]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.schema.metadata :as lib.schema.metadata]
+   [metabase.lib.util :as lib.util]
+   [metabase.util.malli :as mu]))
+
+(defn source-metric
+  "Returns the `:metadata/metric` for the given stage, or nil if this stage is not based on a metric."
+  [query stage-number]
+  (let [stage (lib.util/query-stage query stage-number)]
+    (some->> (:source-card stage) (lib.metadata/metric query))))
+
+(defn join->metric
+  "Given a metadata provider and a join or joinable."
+  [metadata-providerable join-or-joinable]
+  (if (= (:lib/type join-or-joinable) :mbql/join)
+    (when-let [card-id (-> join-or-joinable :stages first :source-card)]
+      (when-let [card (lib.metadata/metric metadata-providerable card-id)]
+        (assoc card ::lib.join/join-alias (:alias join-or-joinable))))
+    ;; :metadata/card or :metadata/metric; look it up by ID.
+    (lib.metadata/metric metadata-providerable (:id join-or-joinable))))
```

Comment:
I think if join is a table, this could have id conflicts (report_card 1, db_table 1)

Reply:
Ah, the later PR #42222 ran into this and I fixed it to be a `case` statement. I could shuffle that code into this first PR if you think it's worth the effort.